src/napari/utils/_magicgui.py

```diff
@@ -432,7 +432,11 @@ def get_layers(gui: CategoricalWidget) -> list[Layer]:
 
     """
     if viewer := find_viewer_ancestor(gui.native):
-        return [x for x in viewer.layers if isinstance(x, gui.annotation)]
+        layers = [x for x in viewer.layers if isinstance(x, gui.annotation)]
+        for layer in layers:
+            layer.events.name.connect(gui.reset_choices)
```

Comment:
These events should be disconnected whenever a layer is removed. As it, if a layer is removed from the viewer but kept around, and then its name is changed, this will cause an update. Probably not a huge deal in most cases, but potentially problematic in terms of holding references around which may cause memory leaks.

Reply:
That's also a good idea, but they serve two different purposes. That one ensures that there is no duplicate connection (so you don't update the gui twice), but what I'm suggesting is to disconnect the layer when it's removed.

Ideally we would have a way to connect to a `layerlist` event that is fired whenever one of its elements changes names, but we don't have this kind of nested machinery yet.

So basically right now we'd need to keep track of the previous layers, and detach callbacks when the layerlist changes. Maybe you can connect to the `removing` event of layerlist and when that fires you can disconnect the layer at that index before it gets removed?

Reply:
Would you want to also disconnect the data connection to _make_choice_data_setter? Currently that doesn't appear to get disconnected anywhere.
I wonder if it's easier to get the layer list to re-emit layer.name? or is having the LayerList listen for the Layer verboten?